haskellPackages.hcoord: fix build

erictapen opened the pull request:
https://github.com/danfran/hcoord/pull/8/ increases version bounds for hcoord. For some reason the build fails when the change is applied with appendPatch, so let's use my branch. The Git tag says version v1.0.0.0 is the same rev as master, so there should be no difference between my branch and master+appendPatch, but however.

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

erictapen committed:
haskellPackages.hcoord: fix build

Increase version bounds for a hcoord dependency [0]. Also disable
checks, as upstream doesn't include the necessary files in the release
tarball [1].

[0] https://github.com/danfran/hcoord/pull/8/
[1] https://github.com/danfran/hcoord/issues/9

cdepillabout commented:
@erictapen Thanks for trying to get this fixed.

Two things:

1. Can you send these Haskell-related updates to the `haskell-updates` branch, instead of `master`?
2. Normally we prefer to wait for upstream to just make a new release fixing these types of problems, since it is less work for us.  Can you wait a week, and if upstream doesn't respond or create a new release, then we can come back to this PR.  Also, in the future, we prefer to call `fetchpatch` on the commit or PR responsible, rather than just updating the underlying source.  There should some examples of using `fetchpatch` for this in the configuration.nix file you can copy.

erictapen commented:
1. Right, forget this everytime.
2. I explained in the commit message and PR text why I'd rather use `overrideSrc`.

Alright, let's wait a while and see how upstream responds.

peti commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -1461,4 +1461,15 @@ self: super: {
 
+  # Remove when https://github.com/danfran/hcoord/pull/8 is merged and a new
+  # release is published to Hackage.
+  hcoord = overrideSrc super.hcoord ({
```
Please don't override the source for the entire package. Instead, please apply the necessary changes as patches to the official release tarball. This approach has the advantage that the patches will *fail* when a new version comes out and this is a good reminder for us to remove the override.

peti commented:
Ping? Will you address the comments?

erictapen commented:
I adressed your comments and rebased to latest `haskell-updates`. Also updated the commit message.

cdepillabout commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -1452,4 +1452,24 @@ self: super: {
+  hcoord = overrideCabal super.hcoord (drv: {
+    # We need to use the git tag, as hackage tarball doesn't contain files necessary for checks:
+    # https://github.com/danfran/hcoord/issues/9
+    src = pkgs.fetchFromGitHub {
+      owner = "danfran";
+      repo = drv.pname;
+      rev = "v1.0.0.0";
+      sha256 = "16a7q2nz0k0z96yyyc9s5pblxaznb9mqj94k8md2a51f36f1vaw4";
+    };
```
It would probably be easier to just mark this package as `dontCheck`, rather than overriding the upstream source.

The problem with overriding the upstream source is that we will have to remember to undo it at some point.  If we don't undo it, we will continue getting the old version.

erictapen commented:
@cdepillabout Mh I get it. But I don't like disabling tests when they could be easily run. How about this:

```nix
    src = pkgs.fetchFromGitHub {
      owner = "danfran";
      repo = drv.pname;
      rev = "v${drv.version}";
      sha256 = "16a7q2nz0k0z96yyyc9s5pblxaznb9mqj94k8md2a51f36f1vaw4";
    }
```

By using `drv.version` we guarantee to use the corresponding source to the Hackage release. If upstream forgets to create the tag, we get notified. We won't get notified when upstream fixes https://github.com/danfran/hcoord/issues/9, but the tests would still run and we won't get notified either when using `dontCheck`.

cdepillabout commented:
I think this still suffers from the fixed-output derivation problem, and we still don't want to override the source of the upstream tarball.

> But it would've helped me to see my reasoning stated above being addressed

From your original PR message, it seemed like your reasoning was the following:

> For some reason the build fails when the change is applied with appendPatch

The "for some reason" part doesn't sound like a valid reason not to use `appendPatch`.  If you don't understand why something doesn't work, definitely freel free to ask us for advice how to get it to work.  (A detailed description of what you tried and what error messages you get definitely makes it easier to help you as well!)

erictapen commented:
@cdepillabout Please note that I wasn't criticizing your rejection of my change, but the way you communicated it first in https://github.com/NixOS/nixpkgs/pull/93469#issuecomment-660643498.

I force-pushed and addressed your concerns.

maralorn commented:
@GrahamcOfBorg build haskellPackages.hcoord

maralorn commented:
@GrahamcOfBorg build haskellPackages.hcoord

maralorn commented:
Tested locally, works fin.

maralorn commented:
Thx, for the PR!
